packages/plugin-deduplicate/sources/commands/deduplicate.ts

```diff
@@ -0,0 +1,78 @@
+import {BaseCommand}                                    from '@berry/cli';
+import {Configuration, LocatorHash, Project, IdentHash} from '@berry/core';
+import {structUtils}                                    from '@berry/core';
+import {Command}                                        from 'clipanion';
+import * as semver                                      from 'semver';
+
+
+// eslint-disable-next-line arca/no-default-export
+export default class DeduplicateCommand extends BaseCommand {
+  static usage = Command.Usage({
+    category: `Workspace-related commands`,
+    description: `Reduces dependencies with overlapping ranges to a minimal set of packages`,
+    details: `https://github.com/atlassian/yarn-deduplicate for yarn v2`,
+    examples: [],
+  });
+
+  @Command.Path(`deduplicate`)
+  async execute() {
+    const configuration = await Configuration.find(this.context.cwd, this.context.plugins);
+    const {project} = await Project.find(configuration, this.context.cwd);
+
+    const locatorsByIdent: Map<IdentHash, Set<LocatorHash>> = new Map();
+    for (const [descriptorHash, locatorHash] of project.storedResolutions.entries()) {
+      const value = locatorHash;
+      const descriptor = project.storedDescriptors.get(descriptorHash)!
+      const key = descriptor.identHash;
+
+      const descriptors = locatorsByIdent.get(key);
+      if (descriptors === undefined) {
+        locatorsByIdent.set(key, new Set([value]));
+      } else {
+        locatorsByIdent.set(key, descriptors.add(value));
+      }
+    }
+
+    for (const descriptorHash of project.storedResolutions.keys()) {
+      const descriptor = project.storedDescriptors.get(descriptorHash)!
+      const locatorHashes = locatorsByIdent.get(descriptor.identHash)!;
+
+      if (locatorHashes !== undefined && locatorHashes.size > 1) {
+        const candidates = Array.from(locatorHashes).filter(candidateHash => {
+          const pkg  = project.storedPackages.get(candidateHash)!
+
+          if (structUtils.isVirtualLocator(pkg))
```

Comment:
You can call `devirtualizeLocator(pkg)` which will give you the source locator. You should then be able to obtain the source pkg through `project.storedPackages.get(sourceLocator.locatorHash)`

Reply:
Thanks! First implementation didn't know about virtual locators and the persisted lockfile was horribly wrong. Was lucky enough to find this line while looking through the other functions.

Reply:
Hm I just copied it from https://github.com/yarnpkg/berry/blob/694e03756b7017a5718d5cbe5ae49bd6e6d8c27e/packages/yarnpkg-core/sources/Project.ts#L1178 which indicates those aren't part of the lockfile anyway. So there isn't anything meaningful I could do anyway, right? I'll leave this for now since I don't understand it and it adds considerable complexity until I can add a test case.

Reply:
Yes, they aren't persisted in the lockfile but they are hydrated back when initializing the `Project` instance: https://github.com/yarnpkg/berry/blob/e487b7e10c2902f48f1064878d236df88e28aad2/packages/yarnpkg-core/sources/Project.ts#L98-L108

Reply:
Comparing the devirtualized packages now.